index.js

```diff
@@ -148,72 +158,75 @@ export default class Delaunator {
+            while (q = hullNext[e], !orient(x, y, coords[2 * e], coords[2 * e + 1], coords[2 * q], coords[2 * q + 1])) {
+                e = q;
                 if (e === start) {
-                    e = null;
+                    e = -1;
                     break;
                 }
             }
-            // likely a near-duplicate point; skip it
-            if (!e) continue;
-
-            const walkBack = e === start;
+            if (e === -1) continue; // likely a near-duplicate point; skip it
 
             // add the first triangle from the point
-            let t = this._addTriangle(e.i, i, e.next.i, -1, -1, e.t);
-
-            e.t = t; // keep track of boundary triangles on the hull
-            e = insertNode(coords, i, e);
+            let t = this._addTriangle(e, i, hullNext[e], -1, -1, hullTri[e]);
 
             // recursively flip triangles from the point until they satisfy the Delaunay condition
-            e.t = this._legalize(t + 2);
+            hullTri[i] = this._legalize(t + 2);
+            hullTri[e] = t; // keep track of boundary triangles on the hull
+            hullSize++;
 
             // walk forward through the hull, adding more triangles and flipping recursively
-            let q = e.next;
-            while (orient(x, y, q.x, q.y, q.next.x, q.next.y)) {
-                t = this._addTriangle(q.i, i, q.next.i, q.prev.t, -1, q.t);
-                q.prev.t = this._legalize(t + 2);
-                this.hull = removeNode(q);
-                q = q.next;
+            let n = hullNext[e];
+            while (q = hullNext[n], orient(x, y, coords[2 * n], coords[2 * n + 1], coords[2 * q], coords[2 * q + 1])) {
+                t = this._addTriangle(n, i, q, hullTri[i], -1, hullTri[n]);
+                hullTri[i] = this._legalize(t + 2);
+                hullNext[n] = n; // mark as removed
+                hullSize--;
+                n = q;
             }
 
-            if (walkBack) {
-                // walk backward from the other side, adding more triangles and flipping
-                q = e.prev;
-                while (orient(x, y, q.prev.x, q.prev.y, q.x, q.y)) {
-                    t = this._addTriangle(q.prev.i, i, q.i, -1, q.t, q.prev.t);
+            // walk backward from the other side, adding more triangles and flipping
+            if (e === start) {
+                while (q = hullPrev[e], orient(x, y, coords[2 * q], coords[2 * q + 1], coords[2 * e], coords[2 * e + 1])) {
+                    t = this._addTriangle(q, i, e, -1, hullTri[e], hullTri[q]);
                     this._legalize(t + 2);
-                    q.prev.t = t;
-                    this.hull = removeNode(q);
-                    q = q.prev;
+                    hullTri[q] = t;
+                    hullNext[e] = e; // mark as removed
+                    hullSize--;
+                    e = q;
                 }
             }
 
+            // update the hull indices
+            this.hullStart = hullPrev[i] = e;
```

Comment:
It seems that before here the hull can be in a broken state? I'm hitting an infinite loop in legalize. Might be an error I made.

Reply:
@fogleman it shouldn't be... Do you have a minimal test case to check if JS works OK?

Reply:
@fogleman but that's not an equivalent change, right? And might have quite a bit of overhead. Three flat arrays of numbers are much lighter than an array of complex linked list nodes with pointers.

Reply:
It's basically equivalent. I just now made the nodes smaller:

https://github.com/fogleman/delaunay/commit/63f09b4c883d2fa9f74a1de7f0a2f97d340072a4

Reply:
The pointers all point to elements within a single array, if that wasn't clear.